### PR TITLE
feat: add scan function in redis and neighbours function in network topology

### PR DESCRIPTION
**Pull request by fcgxz2003**
Add scan function in cache to get a specified number of matching regexp key. 
Add neighbours function to get the specified number of neighbors for source host.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: scheduler/networktopology/network_topology.go [hunk 1]
```diff
@@ -321,6 +324,39 @@ func (nt *networkTopology) ProbedCount(hostID string) (uint64, error) {
 	return probedCount, nil
 }
 
+// Neighbours gets the specified number of neighbors for root node.
+func (nt *networkTopology) Neighbours(root *resource.Host, number int) ([]*resource.Host, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), contextTimeout)
+	defer cancel()
+
+	var neighbours []*resource.Host
+	networkTopologyKeys, _, err := nt.rdb.Scan(ctx, 0, pkgredis.MakeNetworkTopologyKeyInScheduler(root.ID, "*"), math.MaxInt64).Result()
```
Comment (gaius-qi) [inline scheduler/networktopology/network_topology.go:333]: Please don't request redis directly.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/cache/cache.go [hunk 2]
```diff
@@ -82,6 +84,30 @@ type cache struct {
 	janitor           *janitor
 }
 
+// Scan all items to get a specified number of matching regex keys.
+func (c *cache) Scan(p string, n int) ([]string, error) {
```
Comment (gaius-qi) [inline pkg/cache/cache.go:88]: Please refer to other projects.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -82,6 +84,30 @@ type cache struct {
 	janitor           *janitor
 }
 
+// Scan all items to get a specified number of matching regexp key.
+func (c *cache) Scan(m string, n int) ([]string, error) {
```
Comment (gaius-qi) [inline pkg/cache/cache.go:88]:
```golang
func (c *cache) Scan(m string, n int) ([]string, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	regex, err := regexp.Compile(m)
	if err != nil {
		return nil, err
	}

        keys := make([]string, n)
	for item := range c.items {
		if len(keys) >= n {
			break
		}

		if regex.MatchString(item) {
			keys = append(keys, item)
		}
	}

	return keys, nil
}
```

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/networktopology/network_topology.go [hunk 4]
```diff
@@ -76,6 +76,9 @@ type NetworkTopology interface {
 	// ProbedCount is the number of times the host has been probed.
 	ProbedCount(string) (uint64, error)
 
+	// Neighbours gets the specified number of neighbors for root node.
+	Neighbours(root *resource.Host, number int) ([]*resource.Host, error)
```
Comment (gaius-qi) [inline scheduler/networktopology/network_topology.go:80]: ```Neighbours(*resource.Host, int) ([]*resource.Host, error)```

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/networktopology/network_topology.go [hunk 5]
```diff
@@ -321,6 +324,58 @@ func (nt *networkTopology) ProbedCount(hostID string) (uint64, error) {
...
+	defer cancel()
+
+	networkTopologyKeys, err := nt.cache.Scan(pkgredis.MakeNetworkTopologyKeyInScheduler(root.ID, "*"), number)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we can not get a sufficient number of neighbors from the cache, access redis to get them.
+	if len(networkTopologyKeys) < number {
+		networkTopologyKeys, _, err = nt.rdb.Scan(ctx, 0, pkgredis.MakeNetworkTopologyKeyInScheduler(root.ID, "*"), math.MaxInt64).Result()
```
Comment (gaius-qi) [inline scheduler/networktopology/network_topology.go:339]: `math.MaxInt64` => `const valuse`

File: scheduler/networktopology/network_topology.go [hunk 6]
```diff
@@ -321,6 +324,58 @@ func (nt *networkTopology) ProbedCount(hostID string) (uint64, error) {
 	return probedCount, nil
 }
 
+// Neighbours gets the specified number of neighbors for root node.
+func (nt *networkTopology) Neighbours(root *resource.Host, number int) ([]*resource.Host, error) {
```
Comment (gaius-qi) [inline scheduler/networktopology/network_topology.go:328]: `func (nt *networkTopology) Neighbours(srcHost *resource.Host, number int) ([]*resource.Host, error) {`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/networktopology/network_topology.go [hunk 7]
```diff
@@ -321,6 +324,58 @@ func (nt *networkTopology) ProbedCount(hostID string) (uint64, error) {
...
+				logger.Error(err)
+				continue
+			}
+
+			// Add cache data.
+			nt.cache.Set(networkTopologyKey, networkTopology, nt.config.Cache.TTL)
+		}
+	}
+
+	var neighbours []*resource.Host
```
Comment (gaius-qi) [inline scheduler/networktopology/network_topology.go:356]:
```golang
	neighbours := make([]*resource.Host, n) 
	for _, networkTopologyKey := range networkTopologyKeys {
		if len(neighbours) >= n {
			break
		}

		_, _, _, neighbourID, err := pkgredis.ParseNetworkTopologyKeyInScheduler(networkTopologyKey)
		if err != nil {
			logger.Error(err)
			continue
		}

		neighbour, loaded := nt.resource.HostManager().Load(neighbourID)
		if !loaded {
			logger.Errorf("host %s not found", neighbourID)
			continue
		}
		neighbours = append(neighbours, neighbour)
	}
```

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/networktopology/network_topology.go (hunk 5 above)
Comment (fcgxz2003) [reply scheduler/networktopology/network_topology.go]: define const maxScanCount = math.MaxInt64, refer to https://sourcegraph.com/github.com/cockroachdb/cockroach@04dd7ea2c52c5ff378f6a22f58f59de1835b3feb/-/blob

File: scheduler/networktopology/network_topology.go [hunk 8]
```diff
@@ -47,6 +47,9 @@ const (
 
 	// findProbedCandidateHostsLimit is the limit of find probed candidate hosts.
 	findProbedCandidateHostsLimit = 50
+
+	// maxScanCount is the maximum number of work that should be done at every call "Scan" in order to retrieve elements from redis.
+	maxScanCount = math.MaxInt64
```
Comment (gaius-qi) [inline scheduler/networktopology/network_topology.go:52]: `defaultScanCountLimit = 64`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/cache/cache.go [hunk 9]
```diff
@@ -56,6 +57,7 @@ const (
 )
 
 type Cache interface {
+	Scan(p string, n int) ([]string, error)
```
Comment (MinH-09) [inline pkg/cache/cache.go:60]: Scan(m string, n int) ([]string, error)

Comment (MinH-09) [inline pkg/cache/cache.go]: Scan(m string, n int) ([]string, error)

File: pkg/cache/cache_test.go [hunk 10]
```diff
@@ -151,6 +151,61 @@ func TestStorePointerToStruct(t *testing.T) {
...
+	tc := New(DefaultExpiration, 0)
+	tc.Set(v2, v1, DefaultExpiration)
+	tc.Set(v3, v1, DefaultExpiration)
+	keys, err := tc.Scan("^b", 1)
+	if err != nil {
+		t.Error("Couldn't parse a regular expression and returns")
+	}
+
+	if len(keys) != 1 {
+		t.Error("Invalid number of scaning cache keys")
```
Comment (MinH-09) [inline pkg/cache/cache_test.go:164]: Invalid number

**Review by gaius-qi — CHANGES_REQUESTED** [review]
LGTM

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**